packages/node_modules/@node-red/editor-client/src/js/nodes.js

```diff
@@ -1776,119 +1998,134 @@ RED.nodes = (function() {
         }
-        var removedNodes;
-        if (nodesToReplace.length > 0) {
-            var replaceResult = replaceNodes(nodesToReplace);
-            removedNodes = replaceResult.removedNodes;
+
+        // TODO: check the z of the subflow instance and check _that_ if it exists
+        // TODO: Handled activeWorkspace = 0
+        let activeWorkspace = RED.workspaces.active();
+        const activeSubflow = getSubflow(activeWorkspace);
+        for (const node of originalNodes) {
+            const group = /^subflow:(.+)$/.exec(node.type);
+            if (group) {
+                const subflowId = group[1];
+                if (activeSubflow) {
+                    let error;
+                    if (subflowId === activeSubflow.id) {
+                        error = new Error(RED._("notification.errors.cannotAddSubflowToItself"));
+                    } else if (subflowContains(subflowId, activeSubflow.id)) {
+                        error = new Error(RED._("notification.errors.cannotAddCircularReference"));
+                    }
+                    if (error) {
+                        // TODO: standardise error codes
+                        error.code = "NODE_RED";
+                        throw error;
+                    }
+                }
+            }
         }
 
+        const removedNodes = [];
+        // Now that the user has made his choice, replace the nodes that need to be replaced.
+        if (nodesToReplace.length > 0) {
+            const result = replaceNodes(nodesToReplace);
+            removedNodes.concat(result.removedNodes);
+        }
 
-        var isInitialLoad = false;
+        let isInitialLoad = false;
         if (!initialLoad) {
             isInitialLoad = true;
-            initialLoad = JSON.parse(JSON.stringify(newNodes));
+            initialLoad = JSON.parse(JSON.stringify(originalNodes));
         }
-        var unknownTypes = [];
-        for (i=0;i<newNodes.length;i++) {
-            n = newNodes[i];
-            var id = n.id;
-            // TODO: remove workspace in next release+1
-            if (n.type != "workspace" &&
-                n.type != "tab" &&
-                n.type != "subflow" &&
-                n.type != "group" &&
-                n.type != 'junction' &&
-                !registry.getNodeType(n.type) &&
-                n.type.substring(0,8) != "subflow:" &&
-                unknownTypes.indexOf(n.type)==-1) {
-                    unknownTypes.push(n.type);
-            }
-            if (n.z) {
-                nodeZmap[n.z] = nodeZmap[n.z] || [];
-                nodeZmap[n.z].push(n);
-            } else if (isInitialLoad && n.hasOwnProperty('x') && n.hasOwnProperty('y') && !n.z) {
+
+        const unknownTypes = identifyUnknowType(originalNodes, { emitNotification: isInitialLoad });
+
+        const nodeZmap = {};
+        let recoveryWorkspace = null;
+        let missingWorkspace = null;
+        for (const node of originalNodes) {
```

Comment:
I would add all the `node.z` checks (not L2086 block) here to avoid having to create a workspace/subflow everywhere.

Reply:
I think we can handle the whole `node.z` part in two loops:

- 1st: add all workspaces (tab and subflow) from the import.
- 2nd: from the import, search for each node (all except workspaces) if `node.z` already exists or is in the import. If the workspace does not exist, create it according to needs (missing workspace, recovery workspace and active workspace)

This should avoid having to check `node.z` when copying.